### DISCUSSION TITLE
Determine the preceding element

### DISCUSSION THREAD

**Discussion by sebastian-de** [typst/typst#2645] [Q&A] [answered]
Hi,

At heading level 3 I'd like to use inline headings like with the LaTeX `\paragraph{}` command.

The following mostly does the job:
```typst
#show heading.where(level: 3): it => {
    v(1em)
    box(it.body + [#h(1em)])
}
```

However, I'd like to alter the heigth of the spacing, depending on the preceding element. If the last element is another heading (no matter the level), there should be less space before it.

I get that I have to use `locate()` and `query()` somehow, but their usage is quite confusing to me.

Is something like the code below possible?

```typst
#show heading.where(level: 3): it => {
    locate(loc => {
        let last_element = /* how? */
        if last_element == heading {
            v(1em)
        } else {
            v(1.4em)
        }
    })
    box(it.body + [#h(1em)])
}
```

**Comment by Julian-Wassmann** [comment]
Hi Sebastian,

Instead of `locate()` and `query()`, it might be easier with an rest show rule and iterate over all children.
```
#show: rest => {
  let lastItemIsHeading = false
  for item in rest.children {
    // skip empty items between two headings
    if item == [ ] { continue } // you can check for parbreak() too
    // copy current item for querying later
    let lastItem = item
    // modify item if item is heading level 3
    if item.func() == heading and item.level == 3{
      let vSpace = v(1.4em)
      // smaller vertical space if heading is above
      if lastItemIsHeading {
        vSpace = v(1em)
      }
      // styling the inline heading
      item = [
        #vSpace
        #box([*#item.body*] + h(1em))
      ]
    }
    lastItemIsHeading = lastItem.func() == heading
    // return the (modified) item
    item
  }
}
```
Reproduction URL: https://typst.app/project/r7JARG6rO00CTE1qv2vI5P

Does that meet your requirements?

**Reply by sebastian-de** [reply]
Hi Julian,

thanks a lot for your answer! With your code the inline headings work as desired.
However, the `if item == [ ] { continue }` removes spaces in other parts of the document as well, for example around inline equations and inline raw text.

typst always seems to add `space` between two headings, so I adjusted your suggestion:

```typst
#show: rest => {
  let last_item = none
  for item in rest.children {
    if item.func() == heading and item.level == 3 {
      let vSpace = v(1.4em)
      if last_item == [ ] {
        vSpace = v(0.7em)
      }
      item = [
        #vSpace
        #box([*#item.body*] + h(0.8em))
      ]
    }
    last_item = item
    item
  }
}
```
This feels like a hack, but it works without removing spaces around equations.

https://typst.app/project/rj904yqKYCc0DXDL99YPog

Thanks again for pointing me to the right direction!

**Comment by sebastian-de** [comment, marked as answer]
typst scripting is fun! Using `enumerate()` one can get rid of `last_item`:

```typst
#show: rest => {
  for (n, item) in rest.children.enumerate() {
    if item.func() == heading and item.level == 3 {
      let vSpace = v(1.4em)
      if rest.children.at(n - 1) == [ ] {
        vSpace = v(0.7em)
      }
      item = [
        #vSpace
        #box([*#item.body*] + h(0.8em))
      ]
    }
    item
  }
}
```
